### PR TITLE
Print up to 5 last mock invocations for a wrong oder error

### PR DESCRIPTION
**Pull request by Vampire**
This change is [Reviewable](https://reviewable.io/reviews/spockframework/spock/1093)

### CODE REVIEW

File: spock-core/src/main/java/org/spockframework/mock/runtime/InteractionScope.java [hunk 1]
```diff
@@ -20,17 +20,22 @@
...
  *
  * @author Peter Niederwieser
  */
 public class InteractionScope implements IInteractionScope {
+  private static final int MAX_PREVIOUS_INVOCATIONS = 5;
   private final List<IMockInteraction> interactions = new ArrayList<>();
   private final List<IMockInvocation> unmatchedInvocations = new ArrayList<>();
   private int currentRegistrationZone = 0;
   private int currentExecutionZone = 0;
+  private final Semaphore previousInvocationsUpdateSemaphore = new Semaphore(1);
```
Comment (leonard84) [inline spock-core/src/main/java/org/spockframework/mock/runtime/InteractionScope.java]: I don't think we need that, the MockController is already `synchronized` see org.spockframework.mock.runtime.MockController#handle
`unmatchedInvocations` is unguarded as well, you could also use an `ArrayDeque`

**Comment by Vampire** [comment]
> I don't see any tests for the rendering output

Well, there is none, as for the current output which also is not verified in any way, only that the exception is thrown is validated. :-D

File: spock-core/src/main/java/org/spockframework/mock/runtime/InteractionScope.java (hunk 1 above)
Comment (Vampire) [reply spock-core/src/main/java/org/spockframework/mock/runtime/InteractionScope.java]: Hm, I'll have a look again. I just saw it is called on different threads of course, if your SUT works multi threaded. I wasn't aware there is external synchronization

**Comment by Vampire** [comment]
I removed the additional synchronization, rewrote the existing `@FailsWith` tests as `EmbeddedSpecification` and added one that checks that the last 5 invocations are printed properly.

File: spock-core/src/main/java/org/spockframework/mock/runtime/InteractionScope.java [hunk 2]
```diff
@@ -41,9 +45,13 @@ public void addInteraction(final IMockInteraction interaction) {
       public Object accept(IMockInvocation invocation) {
         WrongInvocationOrderError wrongInvocationOrderError = null;
         if (currentExecutionZone > myRegistrationZone) {
-          wrongInvocationOrderError = new WrongInvocationOrderError(decorated, invocation);
+          wrongInvocationOrderError = new WrongInvocationOrderError(decorated, invocation, previousInvocations);
         }
         currentExecutionZone = myRegistrationZone;
+        if (previousInvocations.size() == MAX_PREVIOUS_INVOCATIONS) {
+          previousInvocations.remove();
```
Comment (leonard84) [inline spock-core/src/main/java/org/spockframework/mock/runtime/InteractionScope.java]: If you use `addFirst` and `removeLast` instead then you get the reverse order out of the box, if we change the name to `previousInvocationsInReverseOrder` then there shouldn't be cause for confusion.

File: spock-specs/src/test/groovy/org/spockframework/smoke/mock/OrderedInteractions.groovy [hunk 3]
```diff
@@ -98,44 +165,54 @@ class OrderedInteractions extends Specification {
...
-class RemainingBehaviorSameAsForUnorderedInteractions extends Specification {
-  @FailsWith(TooFewInvocationsError)
+class RemainingBehaviorSameAsForUnorderedInteractions extends EmbeddedSpecification {
   def "too few invocations for one of the when-blocks"() {
-    def list = Mock(List)
-
     when:
-    list.add("foo")
-    list.add("bar")
+    runner.runFeatureBody """
```
Comment (leonard84) [inline spock-specs/src/test/groovy/org/spockframework/smoke/mock/OrderedInteractions.groovy]: I don't think that is necessarily better, when we do no further validation on exception using `@FailsWith` is more succinct and it doesn't require that the code is put into a String. While EmbeddedSpecs are powerful, they add a layer of indirection, that is not necessary here.

File: spock-core/src/main/java/org/spockframework/mock/WrongInvocationOrderError.java [hunk 4]
```diff
@@ -70,6 +80,19 @@ public String getMessage() {
...
     builder.append("Last invocation: ");
     builder.append(lastInvocation);
+    builder.append("\n\n");
+    if (previousInvocations.size() == 1) {
+      builder.append("Previous invocation:\n\t");
+      builder.append(previousInvocations.get(0));
+    } else {
+      builder.append("Previous invocations in reverse order:");
+      List<IMockInvocation> previousInvocations = new ArrayList<>(this.previousInvocations);
+      Collections.reverse(previousInvocations);
```
Comment (leonard84) [inline spock-core/src/main/java/org/spockframework/mock/WrongInvocationOrderError.java]: This should be redundant, see comment below

File: spock-core/src/main/java/org/spockframework/mock/runtime/InteractionScope.java (hunk 2 above)
Comment (Vampire) [reply spock-core/src/main/java/org/spockframework/mock/runtime/InteractionScope.java]: as you wish

File: spock-specs/src/test/groovy/org/spockframework/smoke/mock/OrderedInteractions.groovy (hunk 3 above)
Comment (Vampire) [reply spock-specs/src/test/groovy/org/spockframework/smoke/mock/OrderedInteractions.groovy]: Well, I just did what the documentation says and what seems semantically meaningfuller, which is

> or as a replacement for
> exception conditions in certain corner cases where they cannot be used (like
> specifying the behavior of exception conditions). In all other cases,
> exception conditions are preferable.

But ok, changed it back in the three cases where I don't check the message. :-)

Comment (leonard84) [reply spock-specs/src/test/groovy/org/spockframework/smoke/mock/OrderedInteractions.groovy]: IMO this counts as such a corner case, since you basically have to emulate the whole feature just so you can use the exception condition.

File: spock-core/src/main/java/org/spockframework/mock/runtime/InteractionScope.java [hunk 5]
```diff
@@ -20,17 +20,21 @@
 import org.spockframework.util.ExceptionUtil;
 
 import java.util.*;
+import java.util.concurrent.ArrayBlockingQueue;
```
Comment (leonard84) [inline spock-core/src/main/java/org/spockframework/mock/runtime/InteractionScope.java]: Unused imports

Comment (Vampire) [reply spock-core/src/main/java/org/spockframework/mock/runtime/InteractionScope.java]: Whoops, my bad, fixed.
